Paper: WIP EarthSim: Flexible Environmental Simulation Workflows Entirely Within Jupyter Notebooks

dharhas opened the pull request:
This is not fully complete. Some incomplete sections will be fleshed out over the next few days.

dharhas committed:
add abstract & intro

dharhas committed:
add background.

dharhas committed:
add basic overview of tools

dharhas committed:
add section headers for a rough outline

dharhas committed:
cleanup earthsim definition

dharhas committed:
formatting

dharhas committed:
add placeholder images

jbednar committed:
Anaconda updates

jbednar committed:
Fixed typo

dharhas committed:
Merge pull request #1 from jbednar/jabedits

Anaconda edits

dharhas committed:
add Kim as author and merge in her changes manually

dharhas committed:
add materials, mesh & vel field images

dharhas committed:
add drawing tools

dharhas committed:
formatting

dharhas committed:
fix image

dharhas committed:
Merge pull request #1 from dharhas/earthsim-scipy2018

Earthsim scipy2018

Commit:
kcp_edits

Commit:
update_vel_image

jbednar committed:
Fixed spacing

jbednar committed:
Fleshed out the enhancements sections

deniederhut commented:
Hi everyone! It looks like this paper currently does not build on the preprint server. Can I ask someone to confirm whether this paper builds locally?

jbednar committed:
Clarified novelty

dharhas commented:
@deniederhut I was having trouble building it locally. Haven't tracked down the error yet will look into it.

deniederhut commented:
@dharhas the reviewers will be working off of the PDF on the build server, so it's important that we fix this quickly. I might take a look at your figure captions - this seems to be crashing the build. Also, it looks like someone has removed the header for the references section.

dharhas committed:
fix: labels with underscores break the build

dharhas committed:
Update & fix refereences

dharhas commented:
@deniederhut Thanks for the hint on captions. Looks like you cannot use underscores in label names. It is building now.

dharhas committed:
complete coastline extraction workflow section

deniederhut commented:
😄

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+   proprietary desktop tools that do not interoperate well. In this work, we demonstrate building
+   flexible, lightweight workflows entirely in Jupyter notebooks. We demonstrate these capabilities
+   through examples in hydrology and hydrodynamics using the AdH (Adaptive Hydraulics) and
+   GSSHA (Gridded Surface Subsurface Hydrologic Analysis) simulators. The goal is 
```
`The goal is to provide a set of tools that can easily be reconfigured and repurposed as needed to rapidly solve specific emerging issues.` For which emerging issues? It would be valuable to know which novel areas would be impacted by the usage of such simulation techniques.

dharhas replied:
addressed in commit 583c033

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+The process of building and running environmental simulations using these tools is time consuming, requiring
+a large amount of manual effort and a fair amount of expertise. Typically, the time required to build a 
+reasonable model is measured in months. These workflows support some use cases well, especially multi-year projects 
+where there is often the need for highly accurate, high-resolution physical modeling. But the existing tools and workflows 
```
`But the existing tools and workflows are too heavyweight for other potential applications, such as making short-term operational decisions in novel locations. They also make it difficult to flexibly switch between desktop and remote high-performance-computing (HPC) systems as needed for scaling up and for interactive use.` Are these tools publically available? Is there any reference paper or reports on their implementation? It would be nice to show their limitations explicitly.

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+In this work, we demonstrate building flexible, lightweight workflows entirely in Jupyter notebooks with the aim of
+timely support for operational decisions, providing basic predictions of environmental conditions quickly and flexibly
+for any region of the globe.  For small datasets these notebooks can operate entirely locally, or they can be run with local display and remote computation and storage for larger datasets. We demonstrate these capabilities through examples in hydrology and hydrodynamics using 
+the AdH and GSSHA simulators [cite adh & gssha]. The goal is to provide a set of flexible, high-performance tools that can easily be reconfigured and repurposed 
```
`The goal is to provide a set of flexible, high-performance tools that can easily be reconfigured and repurposed as needed to rapidly solve specific emerging issues.` The same comment made in the abstract section. Also, this can be expanded, as this one is in the main body of the paper.

dharhas replied:
addressed in commit 583c033

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+
+   Example of an unstructured 2D triangular computational mesh of a river that is transected by a roadway embankment with culvert and bridge openings. :label:`mesh`
+
+4. Data gridding: Based on the model specification, any spatially varying material properties, initial conditions and 
```
`Data gridding: Based on the model specification, any spatially varying material properties, initial conditions and time-varying forcing functions (i.e. boundary conditions) are regridded from the original data sources to the computational mesh.` A general comment for this part of the paper, but also on your contributions on regredding(later on the text): Which interpolation algorithms are available? Are all them parallelizable? What is the efficiency of different classes of interpolators in your implementation? When you shift from a parallel execution run in a local desktop to one run in an HPC environment, how the runtimes improve? Are there cases that it doesn't scale well? 

Also how you deal with the scale of analysis? Is always downscaling? If so, any usage of dynamical downscaling? Is there cases that upscaling seems better? How is the support of the tool for these cases?

Roy21py replied:
` In addition, two new open source projects are being released for triangular mesh generation and environmental data access [cite filigree & quest]` It will be good to explain the drawbacks of other python packages that exist. Eg: -pymesh

dharhas replied:
@ricardobarroslourenco Ok I think we have a fundamental miscommunication here. EarthSim is not an end to end solution, but a library of components that are designed to work with well with each other and with the rest of the scientific python ecosystem. Using the existing desktop tools we are limited to whatever interpolation techniques exist inside those desktop tools. Adding new interpolators/functionality requires paying the software vendor to custom code them and add them into the next version. Since EarthSim is a loose library of components, you can use whatever interpolator you can find in the scientific python ecosystem or anything you decide to code yourself. The regridding tools mentioned later in the paper are targeted specifically at providing responsive interactive visualization in the browser... I'll try and clarify this in the text.

dharhas replied:
@Roy21py there is nothing stopping someone using pymesh with the rest of the earthsim tools if it meets the users needs. That said, pymesh wraps Triangle for constrained delauney triangulation which is fairly old software (circa 2005) and does not produce very good quality meshes for hydrodynamic modeling. I'll add something to explain this.

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+5. Simulation: The computational mesh along with the re-gridded data, plus any model parameters (turbulence model, etc.) 
+and forcings required (rainfall, etc.) needed for a specific simulation are written to files formatted for a particular 
+environmental simulation engine. This model is then run with the simulation engine (i.e. AdH, GSSHA). For larger simulations, this 
+is run on an HPC system. 
```
`For larger simulations, this is run on an HPC system.` Which system? What is a general configuration?

Roy21py replied:
`For larger simulations, this is run on an HPC system. ` It will be good to describe "larger simulations" quantitatively. Add examples in terms of - order of magnitude of computation , time of execution, dimesionality of the datasets etc. which exceed the configuration of a local machine and need to use HPC clusters.

dharhas replied:
done

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+operational decisions.  Most of these stages are also locked into particular Windows-based GUI applications that are
+typically tied to execution only on specific desktop machines where they are installed. In most cases, once the model 
+input files are generated, they can be moved to an HPC cluster and run from the command line, but then no GUI is
+available.  This linkage of 
```
`This linkage of computation and visualization can be very problematic, because the local machine may not have enough processing power to simulate the model in a reasonable time, but if the model is simulated remotely, the resulting data files can be too large to be practical to transfer to the local machine for analysis.` This is an important statement. However, it lies in the specifics of HPC environment you are dealing with. Certain HPC environments are not designed for _online interactive_ visual processing, such as high tier CPU Supercomputers. So it would be beneficial to bring context to this statement because large runs often happen in an offline way, and if parameters need to be changed, another run often is necessary.

Roy21py replied:
`This overall pipeline can give very high quality results, but it takes 3-6 months to build and run a model,` Can you cite a source?

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+power to simulate the model in a reasonable time, but if the model is simulated remotely, the resulting data files 
+can be too large to be practical to transfer to the local machine for analysis. 
+
+Moreover, the tools that implement the current workflow are primarily "heavyweight" approaches that encode a wide 
```
It is not true that is not necessary to include assumptions / architectural decisions when running HPC jobs. This could be abstracted to the end user, but such information is indeed necessary. Abstractions may be flexible, either by using various configurations or even with ML-based performance tuning (or setup parametrization), but still, this information needs to be supplied.

Roy21py replied:
`local machine may not have enough processing power to simulate the model in a reasonable time`. Since you are talking about in-memory efficiency of computation, it will be better to explain what is a reasonable timeframe?
Comparing traditional methods with your approach in terms of efficiency later on needs to be quantified in terms of performance improvements.

`resulting data files can be too large to be practical to transfer to the local machine for analysis` - state large in terms of file size, what is the bit rate of file transfer etc. usually if there is a linux scp command on a cluster - the file transfers are pretty fast. Are you talking about some other methods for remote file transfer?

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+EarthSim
+--------
+
+EarthSim is a website and associated GitHub repository that serves two purposes. First, it is a location to work on
```
Please include links to the repo and to the website. If the website is not permanent, include a website citation _(often includes a note: "accessed in dd/mm/yyyy")_

Roy21py replied:
`set of assumptions and architectural decisions` - Please cite a source
`much of the functionality required to implement the above workflow is already available as general-purpose 
+libraries in the Python software ecosystem` - Names of the packages and cite their source as DOI/paper if available.

dharhas replied:
Mostly listed in the EarthSim description section. Not sure that it makes sense to list things like numpy, xarray or the GIS python stack etc.

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+to solve the common Earth Science simulation workflow and visualization problems outlined above. EarthSim aims to demonstrate building
+flexible, lightweight workflows entirely in Jupyter notebooks with the goal of timely support for operational 
+decisions, providing basic predictions of environmental conditions quickly and flexibly for any region of the globe. 
+The overall goal is to provide a set of tools that work well together and can easily be reconfigured and repurposed
```
(Same comment of abstract: need to clarify goal)

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+as needed to rapidly solve specific emerging issues.
+
+EarthSim primarily consists of the core PyViz tools (Bokeh, HoloViews, GeoViews, Datashader, and Param) as well as two
+other new open source tools Filigree and Quest. Short descriptions of these tools follow:
```
Short descriptions of tools are ok, but it would be nice to reference them. Is there just the library guides/repos or there are any paper on their implmentation? (for example scikit-learn is cited as this [paper](http://www.jmlr.org/papers/volume12/pedregosa11a/pedregosa11a.pdf))

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+
+**Param** allows the declaration of user-modifiable values called Parameters that are Python attributes extended to have features such as type and range checking, dynamically generated values, documentation strings, and default values. Param allows code to be concise yet robustly validated, while supporting automatic generation of widgets for configuration setting and for controlling visualizatio ...
+
+All of the above tools are fully general, applicable to *any* data-analysis or visualization project, and establish a baseline capability for running analysis and visualization of arbitrarily large datasets locally or remotely, with fully interactive visualization in the browser regardless of dataset size (which is not true of most browser-based approaches).  The other libraries involved are spec ...
```
`establish a baseline capability for running analysis and visualization of arbitrarily large datasets locally or remotely, with fully interactive visualization in the browser regardless of dataset size` 

This is a bold statement, and hard to prove right. Often for visualization on these tools, you need to reduce data by some sort of encoding/compression/modeling method, or subsampling, as well on relying on distributed processing frameworks for data handling. The visualization libraries don't implement that by themselves. Also, always important to remember the issue of what large is: GB? TB? PB? It would be fun to allocate 1 EB of data at once and render it without loss.

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+**Quest** is a library that provides a standard API to search, publish and download data (both geographical and non-geographical) across multiple data sources including both local repositories and web based services. The library also allows provides tools to manipulate and manage the data that the user is working with.
+
+**Filigree** is a library version of the computational mesh generator from Aquaveo's XMS software suite [cite XMS]. It allows for the generation of high quality irregular triangular meshes that conform to the constraints set up by the user.
+
```
(same comment of citation of papers and repos made before)

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+
+.. figure:: images/drawing_tools.png
+
+   Visualization of drawing tools showing drawn polygons, points, paths, and boundary boxes overlaying a web tile service. :label:`drawingtools`
```
What is the `bounding box` bounding? It seems that the polyline is _leaking_ out of the bbox. The upper polyline seems to have two vertexes leaking out too. How are you calculating this convex hull?

An additional comment is regarding coordinate systems, and ellipsoid. How is this defined in all your images, and in the application as a whole?

dharhas replied:
@ricardobarroslourenco The bounding box and polylines are unconnected, it is juts an example of the type of glyphs that can be drawn over an image and then accessed from python or vice versa.

regarding coordinate systems etc. it is whatever you defined. we are using standard python tools like cartopy etc.

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+
+   roi.data
+
+For example, LANDSAT data can then be retrieved for the ROI as:
```
Cool to load Landsat data. Is it possible to render images? If so, how composites are defined? How is the data loaded? Where is it coming from? GDAL also allows this, so how it would be better/different? How about the size/volume limitations of this load?

Roy21py replied:
Is there a function or module that deals with coordinate reference systems? Once you are integrating multiple data source - you might want to bring them into the same CRS for ease of computation.

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+Similar tools allow editing points, polygons, and polylines.
+
+   
+Enhancements: Annotations
```
This is a very cool feature. How are these drawn assets stored? Is it possible to export them as vectorial data (ex.: shp, kml, GeoJSON, etc)? Do they have a coordinate system assigned?

Roy21py replied:
Is there a way to integrate drawing tools with ArcGIS via arcpy?

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+Many of the datasets used in Earth-related workflows come in the form of multidimensional arrays holding values sampled regularly over some portion of the Earth's surface.  These rasters are often very large and thus slow to transfer to a client browser, and are often too large for the browser to display at all. To make it feasible to work naturally with this data, efficient regridding routines w ...
+
+The Datashader code does not currently provide reprojection of the data into a different coordinate system when that is needed. A separate implementation using the xESMF library was also developed for GeoViews to address this need and to provide additional Earth-specific interpolation options.  The `geoviews.org website <http://geoviews.org/user_guide/Resampling_Grids.html>`__ explains how to use ...
+
```
(same comment on the previous regrid section. Needs to clarify how this is done, because it is a central feature in such system, where datasets need to be compatibilized to a common scale for data intergration).

dharhas replied:
It actually isn't a central feature, the datashader regridding is specifically only for interactive visualization in the browser. For interpolation of data onto say a computational grid, the user would use whatever algorithm or library that they felt was best suited. We have one inside the Filigree package or you could use Scipy.Interpolate or anything else...

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+-------------------------------------------
+
+Although Earth imaging data is typically measured on a regular grid, how quickly the values change across the Earth's surface is highly non-uniform.  For instance, elevation changes slowly in many regions, but very quickly in others, and thus when simulating phenomena like water runoff it is often necessary to use very high resolution in some locations and relatively sparse sampling in others.  T ...
+
```
Triangulated regular network (TIN) is a classical class of interpolator with [several works published](https://scholar.google.com/scholar?q=Triangular+irregular+networks+(TIN)&hl=pt-BR&as_sdt=0&as_vis=1&oi=scholart). How are you implementing it? Is there any improvement?

Roy21py replied:
`Although Earth imaging data is typically measured on a regular grid, how quickly the values change across the Earth's surface is highly non-uniform.` - Can you cite the sources?

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -0,0 +1,364 @@
+
+Crucially, note that very little of the code involved here is customized for hydrology or geographic applications specifically, which means that the same techniques can be applied to different problems as they arise in practice, even if they require changing the domain-specific assumptions involved. 
+   
+
```
Where are these examples?

Roy21py replied:
Adding a diagram of the entire workflow for earthsim would be expected and reasonable. Also mentioning about the dependencies on several other packages and how Earthsim might be affected with upgrades of individual packages involved might be an issue to look into in the future. Also, for some of the example code - the execution time is pretty slow aroun ~2 mins for a cell to run in local machine. If performance improvement is your  motivation then you need to work to make the code execute faster.

Roy21py commented:
The abstract needs to be more clear about the primary objective of the paper: "multiple proprietary desktop tools that do not interoperate well" can be elaborated a bit more to clearly explain where the interoperability issues pop -up; eg:- performance intensive/writing parsers at each step to make multiple modules talk not each other etc. 
The objective is split between two things - building a Jupyter notebook environment vs improvement to existing python packages & creation of new packages - if improvements and creation of new packages were a by product of the main goal - explain clearly where the main motivation of the paper lies instead of focusing on the steps taken to achieve your main goal.

deniederhut commented:
Hi @dharhas ! How are the revisions coming along?

dharhas commented:
@deniederhut 

I've been on vacation overseas. Just got back. We will start addressing the review comments this week.

dharhas committed:
Add example of emerging issues

dharhas committed:
add CMB citation

dharhas committed:
Add additional motivation to intro

sdc50 committed:
* Quest code update
* GSSHA workflow briefly described
* minor edits

dharhas committed:
Tweak abstract and intro

dharhas committed:
Add HPC details

dharhas committed:
clarify EarthSim description

dharhas committed:
Merge branch 'earthsim-paper' into earthsim-paper-sdc

dharhas committed:
Merge pull request #13 from erdc/earthsim-paper-sdc

sdc edits

dharhas committed:
Add clarification on how datashader workds with large datasets

sdc50 committed:
* updates to quest code example with images

sdc50 committed:
add new image

sdc50 committed:
Merge branch 'earthsim-paper' into earthsim-paper-sdc

# Conflicts:
#	papers/dharhas_pothina/dharhas_pothina.rst

sdc50 committed:
fix citations

sdc50 committed:
one more citation fix

dharhas committed:
Merge pull request #14 from erdc/earthsim-paper-sdc

Earthsim paper sdc

dharhas committed:
Add AdH Dambreak example.

dharhas committed:
fix image name

dharhas committed:
add gssha workflow link

dharhas committed:
update last section

dharhas committed:
format fixes

dharhas committed:
fix reference

dharhas commented:
@deniederhut we have a code block that is bleeding into the second column, do you know how to fix the formatting on that?

deniederhut commented:
You'll have to do the line wrapping by hand. Code blocks will respect your
placement of newlines.

On Fri, Jul 6, 2018, 08:20 Dharhas Pothina <notifications@github.com> wrote:

> @deniederhut <https://github.com/deniederhut> we have a code block that
> is bleeding into the second column, do you know how to fix the formatting
> on that?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/415#issuecomment-403032244>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGjNrPJ1gwIjWrq6VG9jFxgXp4K8uLXwks5uD2Q5gaJpZM4ULjyB>
> .
>

dharhas committed:
fix code block formatting

dharhas committed:
improve codeblock indentation

dharhas committed:
add missing figure

dharhas committed:
fix figure references

dharhas committed:
Edits from Kim Peavey

dharhas committed:
fix Hines reference

dharhas committed:
fix reference

dharhas committed:
fix fig reference

dharhas committed:
fix bullet list formatting

dharhas commented:
@deniederhut I think we have addressed the reviewer comments. A couple, I don't think were directly relevant since they seemed to assume that EarthSim was a standalone product whereas it is actually a loose collection of tools. I've tried to clarify that in the description. If there is anything we have missed, please let us know.

deniederhut commented:
Dear @Roy21py and @ricardobarroslourenco , do you feel this paper is now ready for inclusion in the proceedings?

ricardobarroslourenco commented on papers/dharhas_pothina/dharhas_pothina.rst:
```diff
@@ -215,7 +215,13 @@ other new open source tools Filigree and Quest. Short descriptions of these tool
 
-All of the above tools are fully general, applicable to *any* data-analysis or visualization project, and establish a baseline capability for running analysis and visualization of arbitrarily large datasets locally or remotely, with fully interactive visualization in the browser regardless of dataset size (which is not true of most browser-based approaches).  The other libraries involved are spec ...
+All of the above tools are fully general, applicable to *any* data-analysis or visualization project, and establish a baseline capability for running analysis and visualization of arbitrarily large datasets locally or remotely, with fully interactive visualization in the browser regardless of dataset size (which is not true of most browser-based approaches). 
+The key is concept is that the local client system is will always be cabable of performing the visualization, i.e. can deliver it to the user in a browser, regardless of the dataset size.  The assumption is that the remote server will be able to handle the datasets, but because datashader is based on the dask parallel library, it is possible to  assemble a remote system out of as many nodes as re ...
```
`the local client system is will always be cabable of performing the visualization` needs to fix spelling.

dharhas replied:
fixed.

ricardobarroslourenco commented:
@deniederhut I think that it is ready to go. @dharhas I've just put a comment on a spelling change I've caught, but I liked that you addressed all questions. Seems a very nice work :)

dharhas committed:
fix typo

Roy21py commented:
@deniederhut : Most of the comments have been addressed by @dharhas ..so I guess its good to go!
